documentation/External-Grafana/ReadMe.md

```diff
@@ -0,0 +1,68 @@
+This includes design elements to support the following with Grafana operator :
+- Ability to define an external Grafana instance as Grafana source to Grafana operator
+- Ability to create Grafana dashboards on a remote Grafana Instance.
+- Ability to add external cloud data sources to an external Grafana instance.
+
+## Info
+
+status: Draft
+
+## Motivation
+
+Cloud providers have started providing managed remote grafana services which decouples the responsibilities of managing a grafana instanaces from ops personas and kubernetes environment. 
+
+Currently Grafana operator has an integration to add external data sources as a data source to Grafana instances hosted in a kubernetes environment. As more customers starting to use external grafana services, expanding the Grafana operator to support remote Grafana instances becomes inevitable. Adding ability to integrate with external grafana services, adding data sources, creating dashboards and alerting on a remote Grafana instances offloads responsibilities of managing a grafana instanaces from ops personas which helps them to focus on developing the features required for their business. This helps the customer teams to move from self managed Grafana instance on their Kubernetes environments to Pay as you go model on Grafana instances provided by providers.
+
+## Verification
+
+- Create integration tests for adding keys for remote Grafana instance.
+- Create integration tests to create Grafana dashboards on a remote Grafana Instance.
+- Create integration tests to add cloud data sources to remote Grafana instance.
+
+## Current
+
+Currently the grafana operator supports the following for only self managed Grafana Instance :
+- Adding remote remote cloud data sources.
+- Creating Dashboards.
+- Setting up alerting.
+- And Many more.
+
+## Proposal
+
+In short the proposal in this document is about enhancing the Grafana Operator to support the integration to managed grafana services. We would need to enhance the current version of Grafana Operator to support the following :
+
+- Defining external Grafana instance as Grafana source to Grafana operator.
+
+### Defining external Grafana instance as Grafana source to Grafana operator.
+
+Today Grafana operator supports self managed grafana instance as a Grafana source to Grafana operator. `Grafana` CRD of the grafana operator should be enhanced to integrate with external grafana instance.
+
+> CRD `Grafana` will look like. In this sample design, `url` should be preloaded in a ConfigMap and `grafana_api_key` should be loaded to a Secret. Choice of using external secrets or loading secrets manually is end user responsibility :
+
+```.yaml
+apiVersion: grafana.integreatly.org/v1beta1
+kind: Grafana
+metadata:
+  annotations:
+  labels:
+    app: grafana
+  name: grafana
+  namespace: grafana-operator-system
+spec:
+  external:
+    url: <type ConfigMapSelector>
+    grafana_api_key: <type SecretKeySelector>
```

Comment:
Hmm you have a good point about using grafana_api_key.
First of all I think api_key is a better variable name. I at least think assumed since we are in the grafana instance config.

But the second question is how do we want to login to external grafana instances.
A best practice is decadently to use a API key instead of doing

``` 
    admin_username: <type SecretKeySelector>
    admin_password: <type SecretKeySelector>
``` 

The question is if we want to support the other possible solution as well?
Or we can just say that any external grafana instance should always be using a API key.
Or we could support both by using `username` & `password`.
And if you are using an API key you could set username as empty or something like that, i have seen similar solutions on a few API:s through out the years.

But it's not super straight forward and we would have to do allot of docs.
What do you think @elamaran11  and @pb82 ?

Reply:
HI @NissesSenap IMO we should support both authentication mechanisms. `API_KEY` and `username`, `password` so that the user has a flexbility to choose the same accordingly to their needs. We should support separate paremeters for `API_KEY` and `username`, `password`. If `API_KEY` is passed that should take precedence over `username`, `password`.

Reply:
Yeah it's probably easier then doing some strange workaround supporting two in one.

Please describe all of the three options in the document and write somewhere that API key should take precedence over username, password. 
Then we don't forget this during the implementation :)

Reply:
@NissesSenap Added, Please check. I have covered all 3 variants of options.

Reply:
Yes, agree to support both. So basically:

```
spec:
  external:
      admin_username: <type SecretKeySelector>
      admin_password: <type SecretKeySelector>
      admin_api_key:  <type SecretKeySelector>
```

right? And if the API key is provided, the othe two fields are ignored (if provided).

Reply:
@pb82 Thats right. If API Key is provided, that takes precedence over U/P. But also remember `url` should also a mandatory field for external grafana instance.

Reply:
@elamaran11 right, I forgot to include the url.